jobs/AnalyzeResults.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace MediaSearchSignalTest\Jobs;
+
+use mysqli;
+
+class AnalyzeResults {
+
+    private $db;
+    private $searchId;
+    private $out;
+
+    public function __construct( array $config, int $searchId ) {
+        $this->db = new mysqli( $config['db']['host'], $config['client']['user'],
+            $config['client']['password'], $config['db']['dbname'] );
+        if ( $this->db->connect_error ) {
+            die('DB connection Error (' . $this->db->connect_errno . ') '
+                . $this->db->connect_error);
+        }
+        $this->out = fopen(
+            __DIR__ . '/../out/AnalyzeResults_' . $searchId . '.csv',
+            'w'
+        );
+        $this->searchId = $searchId;
+    }
+
+    public function __destruct() {
+        fclose( $this->out );
+    }
+
+    public function run() {
+        $resultsets = $this->db->query(
+            'select id,term from resultset where searchId = ' . intval( $this->searchId )
+        )->fetch_all( MYSQLI_ASSOC );
+        if ( count( $resultsets ) == 0 ) {
+            die( "ERROR: no results found with for search id " . $this->searchId . "\n");
+        }
+
+        fwrite(
+            $this->out,
+            "Term,f1Score\n"
+        );
+
+        $f1Scores = [];
+        foreach ( $resultsets as $resultset ) {
+            $f1Scores[$resultset['term']] = $this->calculateF1Score(
+                $resultset['id'], $resultset['term'] );
+            fwrite( $this->out,
+                $resultset['term'] . "," . $f1Scores[$resultset['term']] . "\n" );
+        }
+        fwrite(
+            $this->out,
+            "ARITHMETIC MEAN," .
+            array_sum( $f1Scores ) / count( $resultsets) .
+            "\n"
+        );
+    }
+
+    private function calculateF1Score( int $resultsetId, string $searchTerm ) : float {
+        $knownGoodImageCount = $this->db->query(
+            'select count(*) as count from results_by_component where ' .
+            'term = "' . $this->db->real_escape_string( trim( $searchTerm ) ) . '" and ' .
+            'rating=1'
+        )->fetch_object()->count;
+        $foundGoodImageCount = $this->db->query(
+            'select count(*) as count from labeledResult where ' .
+            'resultsetId = "' . intval( $resultsetId ) . '" and ' .
+            'rating=1'
+        )->fetch_object()->count;
+        if ( $foundGoodImageCount == 0 ) {
+            return 0;
+        }
+        $foundBadImageCount = $this->db->query(
+            'select count(*) as count from labeledResult where ' .
+            'resultsetId = "' . intval( $resultsetId ) . '" and ' .
+            'rating<1'
+        )->fetch_object()->count;
+
+        $precision = $foundGoodImageCount / ( $foundGoodImageCount + $foundBadImageCount );
+        $recall = $foundGoodImageCount / $knownGoodImageCount;
```

Comment:
I think this calculation for $recall is incorrect.

Recall, AIUI, is basically:
<amount of known good results returned by search> / <total amount of documents known to be good>

$knownGoodImageCount is an accurate reflection of all the data that we know to be good, but $foundGoodImageCount is only a subset, limited by the amount of data that we requested in the first place.

It is very much possible that the search algorithm actually would find the known good document, but at place 105 or 200 (= outside the current limit). Ergo: recall is not only a reflection of the algorithm, it is also affected by how many documents you limit the response by.
By requesting fewer documents, the algorithm will appear to have a worse recall & fscore than it actually has.

Reply:
For this to become valid, the FindLabeledImagesInResults limit should probably be something absurdly large, so that it comes close enough to returning all found results.

Reply:
You are right, but for the purpose of comparing search profiles I don't think it matters